### PR TITLE
Task completion landing page

### PR DESCRIPTION
**Pull request by dianabarsan**
# Description

Adds task completion landing page, which: 
- is only displayed to users that have "can_view_tasks_group" permission. If users are missing the permission, the page is not displayed. Default config is updated to grant this permission to the "chw" role. 
- is only displayed after submitted a task. Routing to the URL manually or using browser history won't re-display the page. 
- depending on the last submitted task, displays tasks that are "Ready" that belong to the same leaf type place. If no tasks are found or a leaf type place was not found in contact lineage, the page is not displayed. 
For example (with default hierarchy), completing a task about a: 
a) person that belongs to a clinic, will display all ready tasks for that clinic
b) clinic, will display all ready tasks for that clinic
c) person that belongs to a health_center, will not display the page
d) health_center, will not display the page
- if the page should be displayed (user has permission, contact belongs to or is leaf type place), but all tasks are completed, the page is not displayed and the snackbar should say "All household tasks completed". 
- when the page is displayed, navigating away from the page to: 
a) another random page (like contacts or reports)
b) a task that is NOT from the same leaf type place
the user should see a modal to confirm navigation, similar to the form edited popup but with different text. The modal should work the same as the form modal, with cancel keeping the user on the page and confirmation navigating away. 
If the user navigates to a task that is from the same leaf type place - whether the click is in the LHS or RHS list, the page should redirect to the task directly, without a modal. 
- the tasks displayed on the page should be identical to the ones displayed on the contact page of the place, including the styling
- For performance reasons (so the user isn't stuck in a loading screen for many seconds while waiting to display tasks), the page uses the tasks that are stored for the tasks list (main component), and cherry picks which tasks to display based on the previous task's contact. This list can become stale between recalculations - this behavior already exists in the tasks list. Also, the list will exclude the last submitted task. The tasks are re-cherry picked every time the main task list is updated. 
- Also for performance reasons, the associated leaf type place is "preloaded" into the store by the task content component, after the form has been rendered (while the database has downtime). 

Recorded telemetry when loading the page: 
- `tasks:group:all-tasks` - total number of tasks for the place, all statuses
- `tasks:group:cancelled` - number of cancelled tasks for the place
- `tasks:group:ready` - number of tasks in "Ready" state 
- `tasks:group:ready:<task_title>` - breakdown of "Ready" tasks by task task_title
- `tasks:group:modal:confirm` - when the user confirms the navigation modal 
- `tasks:group:modal:reject` - when the user rejects the navigation modal 
- `rules-engine:tasks-breakdown:some-contacts` - when fetching tasks counts in the tasks group page

medic/cht-core#5886

# Code review checklist

- [ ] Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- [ ] Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Internationalised: All user facing text
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

**Comment by dianabarsan** [comment]
Hey @latin-panda , could you please review this? Thanks for your time!

**Review by latin-panda — COMMENTED** [review]
Awesome work!! I left a couple of comments inline but all tiny things

File: webapp/src/ts/services/tasks-for-contact.service.ts [hunk 1]
```diff
@@ -80,4 +87,35 @@ export class TasksForContactService {
...
+      return false;
+    }
+
+    const { doc, lineage } = await this.lineageModelGeneratorService.contact(contactId, { hydrate: false });
+    if (!doc || !lineage) {
+      return false;
+    }
+
+    const leafPlaceTypes = await this.leafPlaceTypes$;
+    for (const contact of [doc, ...lineage]) {
```
[the remark this answers is not in the document]
Comment (dianabarsan) [reply webapp/src/ts/services/tasks-for-contact.service.ts:112]: Thanks for noticing!
I used to have an `await` inside the loop, and that only works with `for ... of ...`, but without the await, `find` is better.

File: webapp/src/ts/modules/tasks/tasks-content.component.ts [hunk 2]
```diff
@@ -308,13 +326,14 @@ export class TasksContentComponent implements OnInit, OnDestroy, AfterViewInit {
         console.debug('saved report and associated docs', docs);
         this.globalActions.setSnackbarContent(this.translateService.instant('report.created'));
 
+        this.tasksActions.setLastCompletedTask(this.selectedTask);
```
[the remark this answers is not in the document]
Comment (dianabarsan) [reply webapp/src/ts/modules/tasks/tasks-content.component.ts:329]: We don't, unfortunately. We know that the selected task's action was executed. We don't know what effect it will have, once the tasks are recalculated. 
All we get out of the rules engine is the fresh set of tasks, without knowing which action effected which change.

File: shared-libs/rules-engine/src/provider-wireup.js [hunk 3]
```diff
@@ -86,6 +86,40 @@ module.exports = {
...
+
+    if (!rulesEmitter.isEnabled() || !wireupOptions.enableTasks) {
+      return Promise.resolve(tasksByState);
+    }
+
+    const getTasks = contactIds ? provider.allTaskRowsByOwner(contactIds) : provider.allTaskRows();
+
+    return getTasks.then(taskRows => {
+      taskRows.forEach(({ value: { state } }) => {
+        if (Object.hasOwnProperty.call(tasksByState, state)) {
```
[the remark this answers is not in the document]
Comment (dianabarsan) [reply shared-libs/rules-engine/src/provider-wireup.js:114]: I didn't think so. I think other analytic tools should capture these edge cases, if they ever exist.

File: shared-libs/rules-engine/test/pouchdb-provider.spec.js [hunk 4]
```diff
@@ -249,9 +368,64 @@ describe('pouchdb provider', () => {
...
+        { id: completedTask._id, key: ['owner', 'all', 'patient'], value: { state: 'Completed' } },
+        { id: failedTask._id, key: ['owner', 'all', 'patient'], value: { state: 'Failed' } },
+        { id: cancelledTask._id, key: ['owner', 'all', 'patient'], value: { state: 'Cancelled' } },
+        { id: readyTask._id, key: ['owner', 'all', 'patient'], value: { state: 'Ready' } },
+        { id: draftTask._id, key: ['owner', 'all', 'patient'], value: { state: 'Draft' } },
+        { id: taskOwnedByChtContact._id, key: ['owner', 'all', 'patient'], value: { } },
+      ]);
+    });
+
+    it('should return tasks_by_contact for place', async () => {
```
[the remark this answers is not in the document]
Comment (dianabarsan) [reply shared-libs/rules-engine/test/pouchdb-provider.spec.js:402]: > do we want to consider db exceptions for this one and allTaskRows? thinking

Since these are integration tests (they run against an actual database), it's kind of tricky to simulate database errors, and we don't do it for the other integration tests. I'm adding a test where I pass a malformed parameter and catch the error to show that we expect errors to be thrown.

File: webapp/src/ts/modules/tasks/tasks-content.component.ts [hunk 5]
```diff
@@ -228,6 +233,19 @@ export class TasksContentComponent implements OnInit, OnDestroy, AfterViewInit {
       });
   }
 
+  private preloadTaskGroupContact(action) {
+    this.tasksActions.setTaskGroupContactLoading(true);
+    return this.tasksForContactService
+      .getLeafTypePlaceParent(action?.content?.contact?._id)
```
[the remark this answers is not in the document]
Comment (dianabarsan) [reply webapp/src/ts/modules/tasks/tasks-content.component.ts:239]: I went with this name because I wanted to make it clear why this exists upon first read, that we are pre-loading some contact, no matter what it is, into the store, for another component to consume. 

I purposefully didn't use `get` because I expect `get` to return something (which in this case, it doesn't). Maybe `storeTaskGroupContact` ?

File: webapp/src/ts/modules/tasks/tasks-content.component.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (latin-panda) [reply webapp/src/ts/modules/tasks/tasks-content.component.ts:329]: so `this.tasksActions.setLastCompletedTask` might not be actually completed? Maybe we can named it something different to not confuse the actual state of task 🤔

File: webapp/src/ts/modules/tasks/tasks-content.component.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (latin-panda) [reply webapp/src/ts/modules/tasks/tasks-content.component.ts:239]: I think I'm confused
`getLeafTypePlaceParent` returns an object `{doc: contact, type: type}`, right? It's not storing 🤔

[the remark this answers is not in the document]
Comment (dianabarsan) [reply webapp/src/ts/modules/tasks/tasks-content.component.ts:239]: Yea, but we don't actually return the value, we just store it: 
```
 this.tasksActions.setTaskGroupContactLoading(true);
    return this.tasksForContactService
      .getLeafTypePlaceParent(action?.content?.contact?._id)
      .then(contact => {
        this.tasksActions.setTaskGroupContact(contact);
      })
      .catch(err => {
        console.error('Error when loading task group contact', err);
        this.tasksActions.setTaskGroupContact(null);
      });
```

[the remark this answers is not in the document]
Comment (dianabarsan) [reply webapp/src/ts/modules/tasks/tasks-content.component.ts:239]: Ah! I think I understand now: 
You're not requesting to change `preloadTaskGroupContact`, but `getLeafTypePlaceParent`?

[the remark this answers is not in the document]
Comment (dianabarsan) [reply webapp/src/ts/modules/tasks/tasks-content.component.ts:239]: Sorry for the confusion!

[the remark this answers is not in the document]
Comment (dianabarsan) [reply webapp/src/ts/modules/tasks/tasks-content.component.ts:239]: How about `getLeafPlaceAncestor` ?

File: webapp/src/ts/modules/tasks/tasks-content.component.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (dianabarsan) [reply webapp/src/ts/modules/tasks/tasks-content.component.ts:329]: `setLastSubmittedTask` ?

File: webapp/src/ts/modules/tasks/tasks-content.component.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (latin-panda) [reply webapp/src/ts/modules/tasks/tasks-content.component.ts:239]: `getLeafPlaceAncestor` <- I understand this one better 🙂 Thanks!

File: webapp/src/ts/modules/tasks/tasks-content.component.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (latin-panda) [reply webapp/src/ts/modules/tasks/tasks-content.component.ts:329]: yep!

File: webapp/src/ts/modules/tasks/tasks-group.component.ts [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+    this.globalActions.navigationCancel(nextUrl);
+    return false;
+  }
+
+  private getEmissionIdFromUrl(url) {
+    const tree = this.router.parseUrl(url);
+    // https://angular.io/api/router/UrlTree#example
+    const segments = tree.root.children[PRIMARY_OUTLET].segments;
+    if (segments && segments.length === 2 && segments[0]?.path === 'tasks') {
+      return segments[1]?.path;
```
[the remark this answers is not in the document]
Comment (dianabarsan) [reply webapp/src/ts/modules/tasks/tasks-group.component.ts:253]: This works nicely! It required a few more changes, however, to make sure that the routes we try to access have the extras. 
It also makes the unit tests more fragile, because now we have to mock the router's current navigation instead of just parsing a url, so we depend on the e2e test to verify this interaction.

[the remark this answers is not in the document]
Comment (dianabarsan) [reply webapp/src/ts/modules/tasks/tasks-group.component.ts]: Ok, I went with adding a `state` with a `tab` and `id` property, and added it to the task list and content row list item. 
I didn't add the properties to the other lists, but we could add them as we need them.

**Comment by dianabarsan** [comment]
Thanks a lot for the awesome review @latin-panda . I think I addressed all comments. Could you, please, have another look? Thanks!
